packages/sanity/src/desk/comments/src/context/comments/CommentsProvider.tsx

```diff
@@ -0,0 +1,374 @@
+import React, {memo, useCallback, useMemo, useState} from 'react'
+import {orderBy} from 'lodash'
+import {
+  CommentCreatePayload,
+  CommentEditPayload,
+  CommentPostPayload,
+  CommentStatus,
+  CommentThreadItem,
+} from '../../types'
+import {
+  CommentOperationsHookOptions,
+  MentionHookOptions,
+  useCommentOperations,
+  useCommentsEnabled,
+  useCommentsSetup,
+  useMentionOptions,
+} from '../../hooks'
+import {useCommentsStore} from '../../store'
+import {buildCommentThreadItems} from '../../utils/buildCommentThreadItems'
+import {CommentsContext} from './CommentsContext'
+import {CommentsContextValue, SelectedPath} from './types'
+import {getPublishedId, useEditState, useSchema, useCurrentUser, useWorkspace} from 'sanity'
+
+const EMPTY_ARRAY: [] = []
+
+const EMPTY_COMMENTS_DATA = {
+  open: EMPTY_ARRAY,
+  resolved: EMPTY_ARRAY,
+}
+
+interface ThreadItemsByStatus {
+  open: CommentThreadItem[]
+  resolved: CommentThreadItem[]
+}
+
+/**
+ * @beta
+ * @hidden
+ */
+export interface CommentsProviderProps {
+  children: React.ReactNode
+  documentId: string
+  documentType: string
+}
+
+const EMPTY_COMMENTS = {
+  data: EMPTY_COMMENTS_DATA,
+  error: null,
+  loading: false,
+}
+
+const EMPTY_MENTION_OPTIONS = {
+  data: [],
+  error: null,
+  loading: false,
+}
+
+const noop = async () => {
+  await Promise.resolve()
+}
+
+const noopOperation = {
+  execute: noop,
+}
+
+const COMMENTS_DISABLED_CONTEXT: CommentsContextValue = {
+  comments: EMPTY_COMMENTS,
+  create: noopOperation,
+  edit: noopOperation,
+  getComment: () => undefined,
+  getCommentPath: () => null,
+  mentionOptions: EMPTY_MENTION_OPTIONS,
+  remove: noopOperation,
+  selectedPath: null,
+  setSelectedPath: noop,
+  setStatus: noop,
+  status: 'open',
+  update: noopOperation,
```

Comment:
I would love it if these were all separate context providers, so we can avoid building up the same kind of perf problems we've seen in the `DocumentPaneContext`.
It's always cheaper in React to have many many many context providers in a tree, where you only pay for the providers that are only subscribed, and the `useContext` only cause re-renders in components that need them to re-render.
Compared to super contexts where any change to any of the properties inside the `CommentsContext.Provider` `value` prop will cause expensive re-renders 😅

Reply:
Thanks for the feedback! @stipsan 

Can you clarify what you mean by "_these **were all** separate context providers_"? We could definitely move the `selectedPath` state to its own provider (+ the `mentionOptions`) – but I'm not sure we can move much else to their own providers since they need access to the comments data.

Reply:
Sure they can :)
First of all, a lot of times a state setter doesn't need to know what the current state is. For example `useState` has a state setter that returns the previous value in its updater.
So if this was the case before:
```tsx
import React from 'react'

const StateContext = React.createContext(false)
const SetStateContext = React.createContext(() => {})

export default Provider({children}) {
  const [value, _setValue] = React.useState(false)
  const setValue = React.useCallback(() => {setValue(!value)}, [value])

  return <StateContext.Provider value={value}><SetStateContext.Provider value={setValue}>{children}</SetStateContext.Provider></StateContext.Provider>
}
```
```tsx
import React from 'react'

const StateContext = React.createContext(false)
const SetStateContext = React.createContext(() => {})

export default Provider({children}) {
  const [value, _setValue] = React.useState(false)
  const setValue = React.useCallback(() => {setValue(value => !value)}, [])

  return <StateContext.Provider value={value}><SetStateContext.Provider value={setValue}>{children}</SetStateContext.Provider></StateContext.Provider>
}
```
The Studio codebase also has an example of how to do this with something that uses `useSyncExternalStore` under the hood (`ColorSchemeSetValueContext`, `ColorSchemeValueContext`, and `ColorThemeProvider`) and shows an optimized case.

Secondly, by splitting them up you make hard dependencies explicit and a lot easier to track and understand the flow, and how changes propagate.
And the way React works is that it is able to break up work into units of component renders. It can't partially render a component, it can only partially render a tree. By splitting things up you're allowing React to much better optimize and memoize parts of the tree and reap the full benefits of v18 features. By having super context providers and wrapping them in a `useMemo` is basically the best way to make React run in almost the slowest possible way, and its a lot harder to debug since React devtools only highlight changed props and hooks during profiling, it doesn't highlight which slice of a changed context value prop changed.

This is largely why the React docs when introducing `createContext` showcase this pattern by default: https://react.dev/learn/scaling-up-with-reducer-and-context#moving-all-wiring-into-a-single-file

Reply:
FWIW, we have started splitting up this provider in #4938, but that work has been put on hold due to other priorities.